### PR TITLE
Daily fixes

### PR DESCRIPTION
**Pull request by sinkingsugar**

### CODE REVIEW

**Review by Kryptos-FR — COMMENTED** [review]
A few remarks.

File: src/core/shards/core.hpp [hunk 1]
```diff
@@ -818,12 +818,36 @@ struct SetBase : public VariableBase {
...
   static SHOptionalString help() { return SHCCSTR("Creates a mutable variable and assigns a value to it."); }
 
   static SHOptionalString inputHelp() { return SHCCSTR("Input becomes the value of the variable being created."); }
 
   static SHOptionalString outputHelp() { return SHCCSTR("The input to this shard is passed through as its output."); }
 
+  static inline shards::ParamsInfo setParamsInfo = shards::ParamsInfo(
+      variableParamsInfo,
+      ParamsInfo::Param("Exposed", SHCCSTR("If the variable should be marked as exposed."), CoreInfo::BoolType));
+
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/core/shards/core.hpp:832]: Solves Rare property editor, this makes variables magically editable. Outside Rare right now is just a flag.

File: src/core/foundation.hpp [hunk 2]
```diff
@@ -1272,7 +1277,11 @@ struct CachedStreamBuf : std::streambuf {
 
   void done() { data.push_back('\0'); }
 
-  const char *str() { return &data[0]; }
+  const char *str() {
+    if (data.empty())
+      data.push_back('\0');
+    return &data[0];
+  }
 };
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/core/foundation.hpp:1285]: `const char *` not a const function. `nullptr` is a legal empty C string.

File: src/core/foundation.hpp [hunk 3]
```diff
@@ -371,23 +394,25 @@ struct SHWire : public std::enable_shared_from_this<SHWire> {
...
   uint8_t *stackMem{nullptr};
   size_t stackSize{SH_BASE_STACK_SIZE};
 
-  static std::shared_ptr<SHWire> sharedFromRef(SHWireRef ref) { return *reinterpret_cast<std::shared_ptr<SHWire> *>(ref); }
+  static std::shared_ptr<SHWire> &sharedFromRef(SHWireRef ref) { return *reinterpret_cast<std::shared_ptr<SHWire> *>(ref); }
 
   static void deleteRef(SHWireRef ref) {
     auto pref = reinterpret_cast<std::shared_ptr<SHWire> *>(ref);
+    // if your screen is spammed by the under, don't you dare think of removing this line...
+    // likely a red flag and not a red herring, stop going into kernel land...
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/core/foundation.hpp:415]: lol yeah I almost was tempted so had to write

File: include/ops.hpp [hunk 4]
```diff
@@ -196,8 +196,11 @@ ALWAYS_INLINE inline bool operator==(const SHVar &a, const SHVar &b) {
   case SHType::Color:
     return a.payload.colorValue.r == b.payload.colorValue.r && a.payload.colorValue.g == b.payload.colorValue.g &&
            a.payload.colorValue.b == b.payload.colorValue.b && a.payload.colorValue.a == b.payload.colorValue.a;
-  case SHType::Wire:
-    return a.payload.wireValue == b.payload.wireValue;
+  case SHType::Wire: {
+    auto aWire = reinterpret_cast<const std::shared_ptr<SHWire> *>(a.payload.wireValue);
+    auto bWire = reinterpret_cast<const std::shared_ptr<SHWire> *>(b.payload.wireValue);
+    return *aWire == *bWire;
+  }
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply include/ops.hpp:203]: shards are not shared_ptr behind the scenes so no

File: src/core/foundation.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/core/foundation.hpp:1285]: this is to work with rust btw. UI.Label was crashing. I didn't try `""` though hmm

[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/core/foundation.hpp:1285]: can try it

File: rust/src/shards/gui/widgets/variable.rs [hunk 5]
```diff
@@ -150,3 +170,87 @@ impl Shard for Variable {
...
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, _context: &Context, input: &Var) -> Result<Var, &str> {
+    let input: Table = input.try_into()?;
+    let name: &str = input.get_fast_static("Name\0").try_into()?;
```
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply rust/src/shards/gui/widgets/variable.rs:235]: I mean it's the same

File: src/core/foundation.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sinkingsugar) [reply src/core/foundation.hpp:1285]: seems to work

**Pull request merged**